lib/nodes/addon/components/driver-azure/component.js

```diff
@@ -139,12 +171,118 @@ export default Component.extend(NodeDriver, {
+    if (get(this, 'useAvailabilitySet')) {
+      set(this, 'config.availabilityZone', '');
+    } else {
+      set(this, 'config.availabilitySet', '');
+    }
+  }),
 
-    return regions[environment];
+  vmsWithAcceleratedNetworking: computed('vmSizes', function() {
+    return get(this, 'vmSizes').filter((vmData) => {
+      return vmData.AcceleratedNetworkingSupported;
+    });
+  }),
+
+  vmsWithoutAcceleratedNetworking: computed('vmSizes', function() {
+    return get(this, 'vmSizes').filter((vmData) => {
+      return !vmData.AcceleratedNetworkingSupported;
+    });
   }),
 
+  selectedVmSizeExistsInSelectedRegion: computed('config.{location,size}', 'vmSizes', function() {
+    // If the user selects a region and then a VM size
+    // that does not exist in the region, the list of VM
+    // sizes will update, causing the selected VM size
+    // to disappear. A disappearing VM size seems like a
+    // bad UX, so this value allows the value to be
+    // added to the VM size dropdown, while an error message
+    // indicates that the size is invalid.
+    if (get(this, 'vmSizes').find((size) => {
+      return size.Name === get(this, 'config.size');
+    })) {
+      return true;
+    }
+
+    return false;
+  }),
+
+  vmsWithAvailabilityZones: computed('vmSizes', function() {
+    return get(this, 'vmSizes').filter((vmData) => {
+      return vmData.AvailabilityZones.length > 0;
+    });
+  }),
+
+
+  availabilityZoneChoices: computed('config.{location,size}', 'regions', 'vmSizes', function() {
+    const vmData = get(this, 'vmSizes').find((vm) => {
+      return vm.Name === get(this, 'config.size')
+    })
+
+    if (vmData) {
+      return vmData.AvailabilityZones.map((zone) => {
+        return {
+          name:  zone,
+          value: zone,
+        };
+      });
+    }
+
+    return [];
+  }),
+  sizeChoices: computed('config.size', 'selectedVmSizeExistsInSelectedRegion', 'vmSizes', 'vmsWithAcceleratedNetworking', 'vmsWithoutAcceleratedNetworking', function() {
+    // example vmSize option from backend:
+    // {
+    //   AcceleratedNetworkingSupported: false,
+    //   AvailabilityZones: [],
+    //   Name: "Basic_A0"
+    // }
+    const intl = window.l('service:intl');
+
+    const noAnLabel = intl.t('nodeDriver.azure.size.doesNotSupportAcceleratedNetworking');
+    const withAnLabel = intl.t('nodeDriver.azure.size.supportsAcceleratedNetworking');
+    const vmsWithAn = get(this, 'vmsWithAcceleratedNetworking');
+    const vmsWithoutAn = get(this, 'vmsWithoutAcceleratedNetworking');
+    const out = [{
+      kind:     'group',
+      name:     noAnLabel,
+      value:    noAnLabel,
+      disabled: true
+    }]
+      .concat(vmsWithoutAn)
+      .concat({
+        kind:     'group',
+        name:     withAnLabel,
+        value:    withAnLabel,
+        disabled: true
+      })
+      .concat(vmsWithAn)
+
+    if (!get(this, 'selectedVmSizeExistsInSelectedRegion')) {
+      out.push({
```

Comment:
Unfortunately I don't see _why_, but this doesn't seem to work. When I select a VM size then switch to a region in which the size is not available, I see the warning about the size not being supported, but the dropdown value switches to whatever is first in `sizeChoices` and the invalid size doesn't seem to be in the list (unless I overlooked it, it's a big list)...then when I switch back to the previous region, the dropdown value also switches back to what I had selected.

Reply:
I refactored it and it seems fixed: 

```js
let out = [{
      kind:     'group',
      name:     noAnLabel,
      value:    noAnLabel,
      disabled: true
    }]
      .concat(vmsWithoutAn)
      .concat({
        kind:     'group',
        name:     withAnLabel,
        value:    withAnLabel,
        disabled: true
      })
      .concat(vmsWithAn)

    out = out.map((vmData) => {
      const { Name } = vmData;

      if (vmData.kind === 'group') {
        return vmData;
      }

      return {
        label:    Name,
        value:    Name,
        disabled: vmData.disabled || false,
      };
    });

    if (!get(this, 'selectedVmSizeExistsInSelectedRegion')) {
      return out.concat({
        label:    get(this, 'config.size'),
        value:    get(this, 'config.size'),
        disabled: true
      });
    }

    return out
```
Maybe later this week I'll take some time to figure out why the refactor works but not the original.

Reply:
Oh I think I understand. About array.push:

> It changes the size of the original array and returns a new array (with the new element added) as output.

The way I had it, I conditionally added something to the array, then returned the original array. But when I conditionally added the array item, it created a new array. Whereas the original variable was pointing to the outdated array - the one that didn't have the new item added.

Reply:
.push appends to the array and returns the new length of the array https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/push

concat otoh does return a new array https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/concat

Maybe we're missing something about how ember computed properties work? It's weird that the contents of the dropdown are updated when a new region is selected, and just that option isn't added.

Reply:
I'm pretty sure it's a Javascript thing. What I'm saying is that `const out` was pointing to the memory address of the old array which never got the new item.